xla/backends/gpu/runtime/nvshmem_collective_thunk.h

```diff
@@ -125,6 +125,30 @@ absl::Status IsValidNvshmemOperand(Shape shape, Thunk::Kind reduction_op);
 
 absl::StatusOr<xla::gpu::GpuCollectives*> GetNvshmemCollectivesFromRegistry();
 
+// NvshmemBufferAddresses provides a mechanism to store and retrieve buffer
+// addresses for NVSHMEM put/get operations. This is necessary because NVSHMEM
+// uses one-way put/get operations where both source and destination addresses
+// are required.
+class NvshmemBufferAddresses {
+ public:
+  // Get buffer address for a device
+  absl::StatusOr<void*> GetNvshmemPtr(int device_ordinal);
+
+  // Store buffer address for a device
+  void StoreNvshmemPtr(int device_ordinal, void* buffer_addr);
+
+  // Get the global instance
+  static NvshmemBufferAddresses& GlobalNvshmemBufferAddresses() {
+    static NvshmemBufferAddresses instance;
+    return instance;
```

Comment:
This is not sage initialisation of a static object because it is not trivially destructible: https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables.

Hence why I suggested writing as:
```c++
static NvshmemBufferAddresses* instance = new NvshmemBufferAddresses();
return *instance;
```

Reply:
This change results in a deadlock in the test.

Reply:
Updated